### ISSUE TITLE
Work recovery from failed jobs

### ISSUE THREAD

**Issue by derekchiang** [pachyderm/pachyderm#606]
This has been discussed offline before.  Basically the idea is that, a failed job might have already accomplished a portion of the work, so when we restart the job (speaking of which, there currently isn't a way to restart failed jobs), we'd expect it to pick up from where it left off, instead of doing the entire thing again.

The most intuitive idea is to monitor how much input data has already been consumed, and only feed the restarted job the input data that hasn't been consumed.  The challenge here is that it's unclear how we'd figure out if a piece of data has actually been "consumed", as opposed to simply having been read.  For instance, it's easy to know that a job has read file `foo`.  But how do we know if the job has **acted** on file `foo`?  Might have to introduce some new API.

**Comment by jdoliner** [comment]
I think we're likely going to have to limit ourselves to shard boundaries for work recovery for the reasons you describe here about not really being able to know if a file has been processed or not. That can probably be ok though because we can scale up the sharding to get whatever recovery granularity makes sense.

We've talked about supporting other computation paradigms like classic Map/Reduce. In those tracking completion on a per file basis would totally be doable so maybe it's worth thinking about what we'd need to implement to get that use case working.

**Comment by sjezewski** [comment]
Ran into a similar use case when running through @dwhitena's go compilation pipeline. Basically ... I had parallelism of 64, and doing the rough math I expected the job to complete in about 80 minutes.

However, the input data (list of repos) size is not proportional to computation time. Compiling is CPU bound and some repos seem to take a lot longer. (So the processing is really lumpy)

90 minutes into the job, I have 22 pods left. I'd love to be able to re-distribute the work to the other 42 'idling' VMs, so that we recover some of the total processing time. To do that, we'd have to detect what files have been read as you describe above.

**Comment by sambooo** [comment]
@sjezewski without knowing how long each line/file/block will take to process ahead-of-time, there isn't really any way to know what is left to process at a given time.

Without enforcing some requirements on users regarding filesystem access, we can't just add more data to a running pod since it's entirely possible their code will enumerate files only once and never check for changes. I don't think there is any meaningful way to map file-access info to work-done, since again we don't know the significance of any given read.

So I believe that limits your options to either:

1. Actively scheduling data to containers as they free up,
2. Killing running containers, discarding their outputs entirely and redistributing work.

The first option sounds a lot like it's kubernetes' job which makes me hesitant to try it, but I don't see a reason the second couldn't work while being inefficient.

**Comment by sjezewski** [comment]
I agree its tricky. This is a type of optimization we wouldn't consider for a while, because it probably only yields benefits in niche cases (like the one I listed above).

I do think its possible if the users have the right expectations. Those expectations are basically to re-query to see what files are available between reads.

This prevents the pod that has files 'taken away' from it, from erring when it tries to read a file it may have 'seen' earlier when listing files in a directory.

It also allows pods to handle new files that might appear after initially listing files in a directory.

---

A more broad way to frame this problem is: What recourse we have for 'runaway jobs' ? ... jobs whose processing times are huge outliers. 

One bit of recourse - w the PFS refactor, we'll treat each pod's changes as a separate commit, so if some subset of the pods are long running (as they were above), you should be at least able to stop them, and see which inputs were causing the problem.

**Comment by jdoliner** [comment]
I think we can still do a lot here with job recovery on commit granularity by finding stragglers and restarting them with a higher sharding factor. We'll lose some amount of work along the way but I think the approach can still make things dramatically faster. Especially if we're aggressive about when we call a job a straggler / runaway. For example suppose we have a job with 4 pods running. After a few minutes we notice that 3 of these pods have read 75% of their input data while the fourth has read only 15%. We deem that pod a runaway and so we further shard its input into 8 more containers, those have to start from scratch but they only have `1/32` of the total data so they'll likely finish much earlier than previous run would have.

**Comment by sjezewski** [comment]
There's a separate issue open for straggler detection: https://github.com/pachyderm/pachyderm/issues/442

Meanwhile, in v1.4 the data is distributed at a much finer grain (via datums). I believe that would have helped mitigate the issue I was seeing. Also - PPS is smart enough not to reprocess any datums.

I don't think there is any other todo's for this issue (since we have a separate straggler thread). @derekchiang - care to verify?

**Comment by jdoliner** [comment]
Can confirm, work recovery is as good as it's going to get.
